### PR TITLE
Update ultralytics models

### PR DESCRIPTION
**Pull request by dreadatour**
Update Ultralytics YOLO models to be used in SaaS.

This adds `Segments` model and common `YoloBBox`, `YoloBBoxes`, `YoloPose`, `YoloPoses`, `YoloSegment` and `YoloSegments` models to be used to parse ultralytics results.

### CODE REVIEW

File: src/datachain/lib/models/segment.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            x=[round(coord) for coord in points_x],
+            y=[round(coord) for coord in points_y],
+        )
+
+    @staticmethod
+    def from_dict(points: dict[str, list[float]], title: str = "") -> "Segments":
+        assert len(points) == 2, "Segments coordinates must be a dictionary of 2 lists."
+        assert all(
+            key in points for key in ["x", "y"]
+        ), "Segments coordinates must contain keys 'x' and 'y'."
```
Comment (0x2b3bfa0) [inline src/datachain/lib/models/segment.py:44]: Very minor nitpick: consider using set comparison.

```suggestion
        assert set(points) == {"x", "y"}, "Segments coordinates must contain keys 'x' and 'y'."
```

I doubt it's cleaner or more readable than you approach, though, so feel free to disregard my suggestion.

**Review by 0x2b3bfa0 — APPROVED** [review]

File: src/datachain/lib/models/bbox.py [hunk 2]
```diff
@@ -11,35 +9,101 @@ class BBox(DataModel):
...
+        )
+
+    @staticmethod
+    def from_dict(coords: dict[str, float], title: str = "") -> "BBox":
+        assert (
+            len(coords) == 4
+        ), "Bounding box coordinates must be a dictionary of 4 floats."
+        assert all(
+            key in coords for key in ["x1", "y1", "x2", "y2"]
+        ), "Bounding box coordinates must contain keys with coordinates."
```
Comment (0x2b3bfa0) [inline src/datachain/lib/models/bbox.py:40]:
```suggestion
        assert (
            len(coords) == 4
        ), "Bounding box coordinates must be a dictionary of 4 floats."
        assert set(coords) == {"x1", "y1", "x2", "y2"}, "Bounding box coordinates must contain keys with coordinates."
```

File: src/datachain/lib/models/bbox.py [hunk 3]
```diff
@@ -11,35 +9,101 @@ class BBox(DataModel):
...
-        assert len(bbox) == 4, f"Bounding box must have 4 elements, got f{len(bbox)}"
-        x, y, w, h = bbox
-        return BBox(title=title or "", x1=x, y1=y, x2=x + w, y2=y + h)
+    def from_dict(coords: dict[str, float], title: str = "") -> "OBBox":
+        assert (
+            len(coords) == 8
+        ), "Oriented bounding box coordinates must be a dictionary of 8 floats."
+        assert all(
+            key in coords for key in ["x1", "y1", "x2", "y2", "x3", "y3", "x4", "y4"]
+        ), "Oriented bounding box coordinates must contain keys with coordinates."
```
Comment (0x2b3bfa0) [inline src/datachain/lib/models/bbox.py:93]:
```suggestion
        assert set(coords) == {"x1", "y1", "x2", "y2", "x3", "y3", "x4", "y4"}, "Oriented bounding box coordinates must contain keys with coordinates."
```

File: src/datachain/lib/models/pose.py [hunk 4]
```diff
@@ -8,30 +8,100 @@ class Pose(DataModel):
...
+            x=[round(coord) for coord in points_x],
+            y=[round(coord) for coord in points_y],
+        )
+
+    @staticmethod
+    def from_dict(points: dict[str, list[float]]) -> "Pose":
+        assert len(points) == 2, "Pose coordinates must be a dictionary of 2 lists."
+        assert all(
+            key in points for key in ["x", "y"]
+        ), "Pose coordinates must contain keys 'x' and 'y'."
```
Comment (0x2b3bfa0) [inline src/datachain/lib/models/pose.py:41]:
```suggestion
        assert set(points) == {"x", "y"}, "Pose coordinates must contain keys 'x' and 'y'."
```

File: src/datachain/lib/models/pose.py [hunk 5]
```diff
@@ -8,30 +8,100 @@ class Pose(DataModel):
...
+            y=[round(coord) for coord in points_y],
+            visible=points_v,
+        )
+
+    @staticmethod
+    def from_dict(points: dict[str, list[float]]) -> "Pose3D":
+        assert len(points) == 3, "Pose coordinates must be a list of 3 lists."
+        assert all(
+            key in points for key in ["x", "y", "visible"]
+        ), "Pose coordinates must contain keys 'x', 'y', and 'visible'."
```
Comment (0x2b3bfa0) [inline src/datachain/lib/models/pose.py:94]:
```suggestion
        assert set(points) == {"x", "y", "visible"}, "Pose coordinates must contain keys 'x', 'y', and 'visible'."
```

**Review by 0x2b3bfa0 — COMMENTED** [review]
For consistency with https://github.com/iterative/datachain/pull/592/commits/ab74442723b24bfcebfc4b3b8551b6e3d6706d79

**Comment by dreadatour** [comment]
> ## [Codecov](https://app.codecov.io/gh/iterative/datachain/pull/592?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative) Report
> Attention: Patch coverage is `69.96337%` with `82 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 87.13%. Comparing base [(`059241a`)](https://app.codecov.io/gh/iterative/datachain/commit/059241acfabf270c0e0d3ba128916e0d4b8c4ac8?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative) to head [(`530c14c`)](https://app.codecov.io/gh/iterative/datachain/commit/530c14ca7328e94d9bb338fe06050dec8827b64a?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative).
> > Report is 2 commits behind head on main.
> 
> [Files with missing lines](https://app.codecov.io/gh/iterative/datachain/pull/592?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative)	Patch %	Lines
> [src/datachain/lib/models/ultralytics/bbox.py](https://app.codecov.io/gh/iterative/datachain/pull/592?src=pr&el=tree&filepath=src%2Fdatachain%2Flib%2Fmodels%2Fultralytics%2Fbbox.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative#diff-c3JjL2RhdGFjaGFpbi9saWIvbW9kZWxzL3VsdHJhbHl0aWNzL2Jib3gucHk=)	50.00%	[42 Missing ⚠️ ](https://app.codecov.io/gh/iterative/datachain/pull/592?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative)
> [src/datachain/lib/models/ultralytics/segment.py](https://app.codecov.io/gh/iterative/datachain/pull/592?src=pr&el=tree&filepath=src%2Fdatachain%2Flib%2Fmodels%2Fultralytics%2Fsegment.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative#diff-c3JjL2RhdGFjaGFpbi9saWIvbW9kZWxzL3VsdHJhbHl0aWNzL3NlZ21lbnQucHk=)	54.00%	[23 Missing ⚠️ ](https://app.codecov.io/gh/iterative/datachain/pull/592?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative)
> [src/datachain/lib/models/ultralytics/pose.py](https://app.codecov.io/gh/iterative/datachain/pull/592?src=pr&el=tree&filepath=src%2Fdatachain%2Flib%2Fmodels%2Fultralytics%2Fpose.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative#diff-c3JjL2RhdGFjaGFpbi9saWIvbW9kZWxzL3VsdHJhbHl0aWNzL3Bvc2UucHk=)	69.64%	[17 Missing ⚠️ ](https://app.codecov.io/gh/iterative/datachain/pull/592?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/iterative/datachain/pull/592?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative).

To be honest I am not sure if it worth it now to implement tests for these files. It will require `ultralytics` to be installed and models to be downloaded and will affect test execution time.

I think better idea is to add these models usage into examples?

File: src/datachain/lib/models/ultralytics/segment.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+class YoloSegments(DataModel):
+    """
+    A data model for a list of YOLO segments.
+
+    Attributes:
+        cls (list[int]): The classes of the segments.
+        name (list[str]): The names of the segments.
+        confidence (list[float]): The confidences of the segments.
+        box (list[BBox]): The bounding boxes of the segments.
+        segments (list[Segments]): The segments of the segments.
```
Comment (mattseddon) [inline src/datachain/lib/models/ultralytics/segment.py:90]: this hurts my brain

**Pull request merged**